File: webmvp/src/lib/sessionDisplay.ts
import type { Session } from "@/lib/types";

const TILE_GRADIENTS = [
  "from-emerald-600 to-teal-800",
  "from-violet-600 to-indigo-800",
  "from-rose-600 to-orange-800",
  "from-sky-600 to-blue-800",
  "from-amber-600 to-yellow-800",
  "from-fuchsia-600 to-purple-800",
];

export function sessionTileGradient(sessionId: string): string {
  let hash = 0;
  for (let i = 0; i < sessionId.length; i += 1) {
    hash = (hash + sessionId.charCodeAt(i) * (i + 1)) % TILE_GRADIENTS.length;
  }
  return TILE_GRADIENTS[hash];
}

export function formatSessionDateShort(date: Session["date"]): string {
  return date.toDate().toLocaleDateString(undefined, {
    weekday: "short",
    month: "short",
    day: "numeric",
  });
}

export function formatSessionDateLong(date: Session["date"]): string {
  return date.toDate().toLocaleDateString(undefined, {
    weekday: "long",
    month: "long",
    day: "numeric",
    year: "numeric",
  });
}

export function sessionInitials(title: string): string {
  const words = title.trim().split(/\s+/).filter(Boolean);
  if (words.length === 0) {
    return "♪";
  }
  if (words.length === 1) {
    return words[0].slice(0, 2).toUpperCase();
  }
  return (words[0][0] + words[1][0]).toUpperCase();
}

export function songInitials(title: string): string {
  const words = title.trim().split(/\s+/).filter(Boolean);
  if (words.length === 0) {
    return "♪";
  }
  return words[0].slice(0, 1).toUpperCase();
}

File: mobile/lib/domain/session_display.dart

const _tileGradientKeys = [
  'emerald',
  'violet',
  'rose',
  'sky',
  'amber',
  'fuchsia',
];

/// Stable gradient bucket for playlist tiles (web `sessionTileGradient`).
int sessionTileGradientIndex(String sessionId) {
  var hash = 0;
  for (var i = 0; i < sessionId.length; i++) {
    hash = (hash + sessionId.codeUnitAt(i) * (i + 1)) % _tileGradientKeys.length;
  }
  return hash;
}

String formatSessionDateLong(DateTime date) {
  const weekdays = [
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday',
    'Sunday',
  ];
  const months = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];
  final w = weekdays[date.weekday - 1];
  final m = months[date.month - 1];
  return '$w, $m ${date.day}, ${date.year}';
}

String formatPlaylistDateShort(DateTime date) {
  const weekdays = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  return '${weekdays[date.weekday - 1]}, ${months[date.month - 1]} ${date.day}';
}

String sessionInitials(String title) {
  final words = title.trim().split(RegExp(r'\s+')).where((w) => w.isNotEmpty);
  final list = words.toList();
  if (list.isEmpty) {
    return '♪';
  }
  if (list.length == 1) {
    return list.first.length >= 2
        ? list.first.substring(0, 2).toUpperCase()
        : list.first.toUpperCase();
  }
  return '${list[0][0]}${list[1][0]}'.toUpperCase();
}

String songInitials(String title) {
  final word = title.trim().split(RegExp(r'\s+')).firstWhere(
        (w) => w.isNotEmpty,
        orElse: () => '',
      );
  if (word.isEmpty) {
    return '♪';
  }
  return word[0].toUpperCase();
}
